### ISSUE TITLE
IO - Program BL Heli flashed ESCs via configurator

### ISSUE THREAD

**Issue by MassiveOverkill** [cleanflight/cleanflight#216]
As you can program the ESCs via TX commands I thought it may be easy\convenient to program ESCs via Cleanflight configurator via a modification of the motor tab to send pre-programmed inputs.  The manual for commands is here for Emax motors (please ignore that the manual says 'Simon K'): https://www.massiverc.com/PrestaShop/en/index.php?controller=attachment&id_attachment=31.  For those first reading this request,  it has evolved into not just changing settings but actually flashing ESCs without the need for a separate hardware tool.

**Comment by Jeez-nl** [comment]
Am I correct in thinking this would take the function of the turnigy usb linker to flash ESCs with a bootloader (for example the ones with SimonK as well), while they are connected to the flightcontroller?

**Comment by hydra** [comment]
Probably better to do pass through. Using the ESC tools to configure will require less code and continual maintenance.

**Comment by ledvinap** [comment]
It should be quite easy to modify new softserial code to handle this, only 2 functions need to be changed ...

**Comment by dkisselev** [comment]
@ledvinap What do you mean? The data input would be the USB-connected UART (or whatever serial port is setup as the configurator), and the output would be the standard motor pins.

I imagine it would just be a handler near the top of the standard UART serial code, if (escPassthroughVar == MOTORx) { on serial read, linker write to motorpinx} (escpassthroughvar being something stored in memory so that a reboot resets it),

In the configurator there would be an option to choose which motor to passthrough to, it sends the passthrough cli command and disconnects. Then the user needs to connect with blheli suite or whatever they intend to use.

**Comment by ledvinap** [comment]
Blheli uses nonstandard serial protocol, implementig it from scratch is not that easy in cleanflight, but most of the work is already done for softserial. it will enable reuse of large parts of serial code, with simple serial-to-relay implementation.

**Comment by maymayday90** [comment]
+1 to this feature

**Comment by bulesz** [comment]
+2 ;)

**Comment by automat64** [comment]
This would be an awesome feature

**Comment by rotarypower101** [comment]
This is a really great idea! +3

**Comment by gojibogyo** [comment]
+1 ! 
This would be a really great addition without too much effort and make quad setup so much easier for newcomers as well

**Comment by ctzsnooze** [comment]
yep +1 from me

**Comment by barthchris** [comment]
+11! 

Im just a mere mortal here, I have no coding ability besides 1979 era  TI-99/4A BASIC!  Wish I could contribute, but to those who have the ability I think this would be much used option, .

 Without you guys we wouldn't be enjoying this awesome hobby. Thanks in advance if it does come to fruition. 

Chris

**Comment by hydra** [comment]
This might help - someone port it and create a PR?

https://github.com/multiwii/baseflight/compare/master...DTFUHF:usblinker

**Comment by dkisselev** [comment]
@hydra nice! That looks like exactly what I was thinking as a solution to this, thanks for cherrypicking the specific commits. 

I'm interested in implementing it but as usual time is tough to come by, I'll keep you guys posted if I make any progress.

Interestingly enough it looks like they set theirs up to only work on motor pin 1. If I implement it, I'd add a parameter to let people choose which motor they want to connect to. Also looks like there's no escape sequence, so a way of quitting out of programmer mode without rebooting the whole FC would be good.

**Comment by MassiveOverkill** [comment]
This is great news!

**Comment by jaimevisser** [comment]
Even if a reboot is required to switch back to config mode that would still be awesome! With an escape sequence in a passthrough configuration you run the risk of other software sending that sequence accidentally and messing up an ESC.

**Comment by dkisselev** [comment]
@jaimevisser I figured that there would exist some sort of sequence that linker software would never put out, but that's a good point, it's not worth the risk of bricking ESC's for the tiny convenience of not having to reboot.

**Comment by 4712** [comment]
Anyone working on the linker implementation?

**Comment by jaimevisser** [comment]
Just thought that this kind of pass through could also work for programming a minimosd.

**Comment by MassiveOverkill** [comment]
I'm throwing in a free set of SN20As and a SeriouslyProF3 racing board to
the dev who implements this.


*Romeo aka MassiveOverkill*

*MassiveRC LLC <http://www.massiverc.com/>*

*MassiveRC Shop <http://www.shop.massiverc.com/>*

*massiveoverkill@gmail.com <massiveoverkill@gmail.com>*

*skype: massiveoverkill*

**Comment by hydra** [comment]
To get things started I've imported the usb linker code from the commit mentioned earlier into a new cleanflight branch.

https://github.com/cleanflight/cleanflight/commits/usblinker

**Comment by jaimevisser** [comment]
@hydra yep, but that will only work for simonk based escs. Just checked out cleanflight. It's been a while since I did anything in C so do not expect to much but I must say that the code is very clean and readable. Is there any documentation on the blheli protocol? Anything will help, even the arduino sketch/source would be a great help!

**Comment by jaimevisser** [comment]
Have been searching for any source for the arduino blheli linker but have only been able to find hex files. Asked the BlHeli devs on RCgroups if they could provide source or any other help.

http://www.rcgroups.com/forums/showthread.php?t=2136895&page=247#post31655280

**Comment by jaimevisser** [comment]
BlHeli linker with multiwii integration available at https://github.com/c---/ArduinoUSBLinker/blob/master/ArduinoUSBLinker.ino

**Comment by maymayday90** [comment]
+1 for MinimOSD programming as well (does this include uploading new MWOSD sketches?). Most of the parameters can be changed via the FPV link though, so not as important as the BLHeli support.

Supporting both bootloaders would also be awesome, or maybe BLHeli bootloader could support the Linkers? (As it doesn't currently.)

**Comment by fpvrecursion** [comment]
I've successfully been able to change the BLHeli parameters with AVRDUDE and a USB Linker by changing the EEPROM values. Granted, this is for the SK bootloader.

Here is where you can find the map for the values:
https://github.com/bitdump/BLHeli/blob/master/Atmel/BLHeli.asm#L1268

Mode 2 (multi) starts at line 1268.

**Comment by fpvrecursion** [comment]
Does anyone have access to the source of the 1-wire Arduino hex files in the BLHeliSuite?

**Comment by fpvrecursion** [comment]
Unless someone has made any progress on this, I can start doing some proof-of-concept work using the USB Linker tonight.

**Comment by jaimevisser** [comment]
@fpvrecursion See my previous comment for source of the arduino linker for atmel BlHeli and a link to the rcgroups topic where the BlHeli devs are. I haven't done any significant work on this yet so if you want to give it a shot, be my guest!

**Comment by fpvrecursion** [comment]
@jaimevisser That's fine for the ESCs that have the SK bootloader (which uses the STK500v2 protocol) on them, but ESCs flashed with the BLHeli bootloader use 1-wire for read/write. At least as far as I understand it. I've had success with the SK bootloader, but since BLHeliSuite doesn't have the source to the 1-wire Arduino implementation, I don't know where to start with that.

Hopefully, a majority of people are using BLHeli with the SK bootloader.

**Comment by 4712** [comment]
What is meant with "1-wire Arduino hex files in the BLHeliSuite", this is a simple TX/RX to 1-wire bridge w/o any other functionality...

**Comment by fpvrecursion** [comment]
Right, but how do we read/write to the eeprom with 1-wire? Do you have examples?

**Comment by 4712** [comment]
The original thread for AVRootloader can be found here: http://www.mikrocontroller.net/topic/avr-bootloader-mit-verschluesselung. Unfortunately it is not very rich commented...
As soon if I find the time, I will post a commented version, which also includes the feature "read flash".

The bootloader uses 19200 baud fixed.
You need to send 8* 0x00 + 0x0D+ 'BLHeli'+crc16 and get back '471'+ other bytes.
The easiest is, to look at the log of BLHeliSuite. For the connect procedure set log level to 4.
![blhelisetuplogwindow_21 05 2015_2](https://cloud.githubusercontent.com/assets/1628451/7759502/449c2038-0012-11e5-95c5-fe3460f61e60.png)

**Comment by fpvrecursion** [comment]
@4712 Thanks! That's enough to get me started.

And thanks for all the work you do on BLHeliSuite too!

**Comment by 4712** [comment]
BTW: wouldn't it make more sense to implement the 4w-if protocol? (excluding the SiLabs C2 part)
How much space is available for such a solution?

**Comment by jaimevisser** [comment]
4w-if would be the best option, but from what I can see in my enviroment most people are running Atmel BlHeli and are using arduino 1-wire to flash and configure their ESCs.

**Comment by fpvrecursion** [comment]
If we use 1-wire or the usb linker, we won't have to remove the ESC from the flight controller. Wouldn't you have to solder to the pads on the ESC or use a socket tool for 4w-if?

**Comment by 4712** [comment]
4w-if integrates: 
- 1-Wire protocol for BLheli bootloader (Atmel and Silabs) 
- USB Linker (stk500v2) for SimonK bootloader 
- It does cover SiLabs C2 interface too, which should be removed for the via-FC programming purpose.
4w-if does not cover ISP flashing (socket tool).

Also some code could be removed, which is used to read and write Atmel flash.
What would remain:
1. Read/Write EEprom for Atmel over BLHeli and SK bootloader.
2. Read/Write Flash /Erase page for SiLabs over BLheli bootloader.

**Comment by hydra** [comment]
Lets focus this:

Phase 1 - one wire support for blheli firmware.

Go!

**Comment by fpvrecursion** [comment]
Haha... Alright... I'm on it.

**Comment by 4712** [comment]
As an interface for Cleanflight GUI or as pass through?

**Comment by jaimevisser** [comment]
Passthrough would be easier to start with. Having blheli support in
cleanflight config would make it multiplatform available but integrating it
would be a large effort.

**Comment by 4712** [comment]
Well, a simple pass through is easy done, but the problem is, how to switch it on or off?
I mean, it would be annoying to always have to reboot the FC, if one wants to change the ESC to talk to.

**Comment by fpvrecursion** [comment]
Hmm... Well I already have an external USB linker with STK500v2 already partially working through the GUI. Next step would be to get the linker code inside the STM32 so we can talk to the ESC directly through the FC.

Switching my GUI stuff to 1-wire shouldn't be very difficult, I'm just not very familiar with the 1-wire protocol for talking to the ESC. I'd have to base it largely off of BLHeliSuite logs and 4712's pointer.

**Comment by 4712** [comment]
@fpvrecursion: which GUI?

**Comment by fpvrecursion** [comment]
Cleanflight Configurator

**Comment by hydra** [comment]
@4712 Passthough for now.

**Comment by 4712** [comment]
Domnic, where to start coding and debugging cleanflight?
I managed to compile with cygwin64, but where to jump in now with the 1-wire bridge code?

**Comment by hydra** [comment]
@4712 it sounds like @fpvrecursion is already on the case.  See the developer docs, see `cliGpsPassthough`.  Maybe have a skype or irc chat with @fpvrecursion so you don't duplicate efforts.

**Comment by 4712** [comment]
Ok, I will wait out on the completion.

**Comment by fpvrecursion** [comment]
4712 clearly has more experience with this than I do. I'll gladly yield to him and fill in where he needs me to.

**Comment by fpvrecursion** [comment]
@4712 I'm in IRC now if you need to talk. Please take this over if you're able and I'm here for any help you need.

**Comment by 4712** [comment]
@fpvrecursion: I will need some time for BLHeliSuite; after this (2-3 days) I will have a look at cleanflight and get back with questions.

**Comment by jaimevisser** [comment]
I will be online tomorrow 11am-~1630pm GMT+1 to help on this. Ideas for integration into cleanflight.

Store an int flag in non-volatile memory, let's call this flag `Passthrough`. Passthrough = 0 by default.

In Cleanflight bootup code:
```
read Passthrough into variable p.
write 0 to Passthrough
if p>0 
 start 1-wire code with motor p
else
 normal bootup
```

In the serial port command handling, add a command `blhelimode x` that does the following:
```
if 0 < x && x =< num_motors
 Set Passthrough to x
 reboot
else
 output help text for command to serial
```

This all ensures that the 1-wire mode can be turned on and is gone after a reboot.

**Comment by hydra** [comment]
@jaimevisser the same approach is used in cleanflight with the spektrum bind code.

Anyone considering this approach should investigate how that works and either:
a) use the same approach (store setting in EEPROM and clear it on reboot)
b) use memory (not EEPROM) that persists after a reboot (if ESCs don't need to be powered off)

IMHO It would be preferable to avoid eeprom writes for both solutions and approach B (above) should be used, but this can only work if the ESCs can be programmed after a FC soft reboot instead of a cold boot of everything (FC's and ESCs).

**Comment by hydra** [comment]
Thinking about it more, the spektrum bind stuff needs a cold-boot as the receivers only enter bind mode after cold boot of the receiver.

**Comment by ledvinap** [comment]
@hydra: There should be no need to reboot from CF perspective - just deconfigure PWM outputs and switch to bootloader code.
Cold boot will be necessary if ESC needs power-on to enter boot mode. It would be best to enter bootloader on all escs simultaneously and switch from configurator as needed ...

**Comment by hydra** [comment]
@ledvinap great, thats even better then.  No changes needed to startup code in that case.

**Comment by jaimevisser** [comment]
Most ESCs I know "arm" when they receive a valid pwm/oneshot signal. After that they don't accept bootloader 1-wire as far as I'm aware. You can of course power the Naze from USB, then go into linker mode and then connect the battery (with power limiting bulb if you have one) to get the ESCs connected. If however you have an FC that is powered from the battery there is no way to program the ESCs without disconnecting some wires or rebooting the FC. My advise would be to implement the cold boot scenario as that would work in most cases, thus bringing the easiest and most fool-proof user experience.

**Comment by DarkVegetableMatter** [comment]
How about this:
-Start FC with USB connected, enter CLI and enable 'passthru to ESC X'. User disconnects configurator, connects BLHeli suite
-Connect battery, powers up ESC, BLHeli suite detects ESC and configures it
-Exit BLHeli suite, disconnect battery, disconnect USB
-Rinse and repeat until all ESCs are configed

This scheme doesn't require NVM etc although it is a bit cumbersome

**Comment by 4712** [comment]
I think, the part with entering CLI and enable passthru can be later implemented into BHeliSuite.
Also there is no need to cycle power for the ESC to connect (at least not since the ones flashed with  BS 13.0.0.0).

I managed to install eclipse luna and got cleanflight compiled. 
But I have absolutely no idea where to start in the code.
Btw: is there any chance to get cleanflight work with CoIDE?

**Comment by 4712** [comment]
Because I just do not have enough time to familiarize myself with the architecture of Clean Flight, I have meanwhile released the code for my Atmel solution.

https://github.com/4712/BLHeliSuite/blob/master/Interfaces/Arduino1Wire/Source/Arduino1Wire_C/Arduino1Wire.c

**Comment by nathantsoi** [comment]
I needed to re-configure some of my ESCs and I thought it would be easier to implement this than unplug everything... not sure if that's true, but nonetheless, I'm workin' on it.

This is my first look at Cleanflight. Any comments/suggestions would be great: https://github.com/nathantsoi/cleanflight/pull/1/files since I have no idea if [this strategy](https://github.com/nathantsoi/cleanflight/blob/usblinker/docs/1wire.md) will work.

For example, is there a better way [than this](https://github.com/nathantsoi/cleanflight/pull/1/files#diff-8a3d178072f596f2ddd7157a9b97ccedR57) to get `pin` and `gpio` for the target ESC port, maybe from `timerHardware` in `timer.c`?

At least it compiles with `make clean TARGET=CC3D && make TARGET=CC3D` but with `OPBL=yes` it runs out of flash.

`OPBL=yes` builds without `#define DISPLAY`

I'll flash my CC3D tomorrow and test this out. Assuming I get something working at some point, I'll write some tests, rebase and make a pull request against the main Cleanflight repo.

**Comment by nathantsoi** [comment]
Ok, I bought a dev board and straight up ported @4712's Arduino code, but sadly, it's not working.

The basic idea is to plug a UART/USB adapter into the board somewhere and pass through to the 1 pin on the ESC. Pick the USB/COM option in BlHeli suite and use as normal.

I get Tx/Rx on the UART, but BlHeli thinks it isn't getting a response, so I'm thinking my timing or levels must be off.. but I don't have a scope to check. Maybe it's time to buy one.

It would be awesome if someone wants to take a look at the code and give me some pointers or hook up a STM32F3Discovery to a scope.

You can run just the 1wire passthrough feature using this code:
https://github.com/nathantsoi/stm32-serial-1wire-passthrough, the [readme](https://github.com/nathantsoi/stm32-serial-1wire-passthrough/blob/master/README.md) explains how.

Or run all of Cleanflight from my [fork/branch](https://github.com/nathantsoi/cleanflight/tree/usblinker), go into the console and run `1wire 1` or whatever ESC you have connected and test that way.

Docs are here: https://github.com/nathantsoi/cleanflight/blob/usblinker/docs/1wire.md

**Comment by 4712** [comment]
Code is working fine so far, but exceeds the 128k flash size. So what to disable?
#ifdef USE_SERIAL_1WIRE
#undef USE_FLASHFS 
//#undef DISPLAY //??
#endif
or better alternatives?

**Comment by nathantsoi** [comment]
@4712, building w/ gcc4.8 & flashing this branch https://github.com/nathantsoi/cleanflight/tree/usblinker with `TARGET=NAZE make clean && TARGET=NAZE make` seems to work fine for me. what fork/branch/revision are you building?

btw, what ESCs are you testing with? I still can't get it working with my UBEC20s or blueseries 12a... maybe i'm building the wrong code?

**Comment by nathantsoi** [comment]
If anyone else wants to try this, I've built master (as of right now) + serial1wire:

naze hex: https://www.dropbox.com/s/c4z2tj2q63jz8rv/07.12.2015-usb1wire-cleanflight_NAZE.hex?dl=1
md5: 70f565c7c4e1d781def8e23f71693f7f

usage is described here: https://github.com/nathantsoi/cleanflight/blob/usblinker/docs/1wire.md#usage

on the naze, the onboard USB/serial converter is used, so 1) flash your board 2) open the configurator cli and type `1wire 1` (to configure the first esc) 3) disconnect the configurator 4) switch to blheli and use as normal (after selecting silabs or atmega type `1` (usb passthrough) programmer

also, trying to build the opbl code for the cc3d, im seeing a flow overflow of 36 bytes.

**Comment by 4712** [comment]
Done my test fork here https://github.com/4712/cleanflight/
I'm still noob with Github...
I will check your hex on my Flip board.

**Comment by 4712** [comment]
I've looked at your docu:
"Attach your 1-wire or 4w programmer (made via the BlHeli suite) to the pin specified in the board documentation."

is this simply outdated? or a misunderstanding?

**Comment by nathantsoi** [comment]
Outdated, sry. I'll update those. Directions in the post above are correct.

I will double check my escs. 

Just to be sure. My escs work with the arduino 1wire code you posted, therefore they must have the blheli bootloader, right?

**Comment by 4712** [comment]
The ESC should work with an arduino board flashed as interface for "Arduino BLHeli Bootloader"

**Comment by nathantsoi** [comment]
Ok I made a programmer with 14.0.0.0:

![screen shot 2015-07-12 at 5 57 37 pm](https://cloud.githubusercontent.com/assets/399770/8640980/911fcd8a-28bf-11e5-8898-dfb5fb091885.png)

then i tried to connect using the interface:

![screen shot 2015-07-12 at 5 58 32 pm](https://cloud.githubusercontent.com/assets/399770/8640985/bf2dae9a-28bf-11e5-85e9-91a000d3d289.png)

and it failed:

![screen shot 2015-07-12 at 5 57 01 pm](https://cloud.githubusercontent.com/assets/399770/8640987/de611fcc-28bf-11e5-9ab8-affb949c787c.png)

![screen shot 2015-07-12 at 5 57 27 pm](https://cloud.githubusercontent.com/assets/399770/8640986/cab67472-28bf-11e5-9f5b-83fdd678c10b.png)

then i switched to 13.2.0.3 and it worked:

![screen shot 2015-07-12 at 5 56 33 pm](https://cloud.githubusercontent.com/assets/399770/8640989/e71f6ee8-28bf-11e5-959c-3b073c36130d.png)

do i need to update the bootloader on my escs?

**Comment by 4712** [comment]
Thanks, this seems to be a bug...

**Comment by hydra** [comment]
Great progress guys, keep the testing and feedback coming

**Comment by nathantsoi** [comment]
Awesome, 14.0.0.1 works with the Arduino flashed with BlHeli Bootloader Tool (C/1)

Unfortunately, the naze (flip32 full actually) isn't going. I did not upgrade the bootloader (assuming it's the same as when I flashed via 13.1.0.0), but I upgraded BlHeli on one ESC via the 14.0.0.1 configurator. Both the 13.x and 14.x ESCs behave the same way. 

I flash my flip32, open the cli and send `1wire 1`, connect the UART adapter to my windows xp vm and hit connect with BlHeli. It just hangs here:

![screen shot 2015-07-12 at 10 06 17 pm](https://cloud.githubusercontent.com/assets/399770/8643138/5910bbc8-28e4-11e5-9f58-caecb00df2e9.png)

Here's a video of the whole process: https://www.youtube.com/embed/sWMmjjI_H_U

I was concerned the VM might be causing some problems, so I tried it on my windows 8 laptop as well -- same result.

Any suggestions what to try next?
